### DISCUSSION TITLE
Problem producing an importable extension on macOS ARM64

### DISCUSSION THREAD

**Discussion by lelit** [pypa/cibuildwheel#1733] [Q&A]
Hi, I'm seeking some cue about a linkage problem I do not understand.

 I'm the author of [pglast](https://github.com/lelit/pglast): it basically builds some facilities on top of a [static library containing the PostgreSQL parser](https://github.com/pganalyze/libpg_query).

Recently I added silicon macs to the [build matrix](https://github.com/lelit/pglast/commit/3b291b8174816e6dc095a2a54d092af9ecd66dc8). Since I do not own that kind of machine, and that [I disabled CI tests](https://github.com/lelit/pglast/commit/a1d9efb6facc822c79608a21cf5ddccb809b4121) for that combo, it went unnoticed that, although the process [completed successfully](https://github.com/lelit/pglast/actions/runs/7615284163/job/20739458193), the wheel contains a [broken shared object](https://github.com/lelit/pglast/issues/144).

A friend lent me an hour of his M3 machine, and I found that, apparently, the ARM64 `pglast/parser.cpython-311-darwin.so` was not properly linked against the static `libpg_query.a`, as [even the size suggests](https://github.com/lelit/pglast/issues/144#issuecomment-1912150169): the Intel one is much bigger, and inspecting them with `nm` it actually contains all the expected `_pg_query__XYZ` concrete symbols, missing in the ARM64 one.

I looked around and tried a [different approach](https://github.com/lelit/pglast/commit/975d02ad7fda6fc30b1bdf9524360de5cc9c7903), but even that does not seem to solve the problem.

I wonder if some kind soul around here, more expert than me on that architecture, may shed some light on the problem?

Thanks in advance for any hint!

**Comment by joerick** [comment]
Hi there

> Recently I added silicon macs to the [build matrix](https://github.com/lelit/pglast/commit/3b291b8174816e6dc095a2a54d092af9ecd66dc8).

You might already know this, but this config is not specifying an arm64 machine. This [runner is on x86_64](https://github.com/lelit/pglast/actions/runs/7608532091/job/20717882014#step:6:192), Github hasn't released arm64 runners yet(!). So this is cross-compiling arm64 from x86_64.

> I looked around and tried a [different approach](https://github.com/lelit/pglast/commit/975d02ad7fda6fc30b1bdf9524360de5cc9c7903), but even that does not seem to solve the problem.

An issue here is that you're not detecting the right case in setup.py. Because we're running this build on x86_64, the `platform.mac_ver()` is still x86_64. Take a look at `sysconfig.get_platform()`, that will return the correct result when cross-compiling.

But I suspect the issue is that your dependency, libpg_query, isn't cross-compiling well. cibuildwheel does set the `ARCHFLAGS` variable, but there might need to be some more options fiddling to get it to work.

But it's going to be easier to do it native. [CirrusCI](https://cirrus-ci.org/) does offer (limited) free arm64 CI for open source, perhaps you can try that.

**Reply by lelit** [reply]
Thanks a lot, I will insist in my investigation, and read about CirrusCI.
Maybe you happen to know how easy/possible is it to _simulate_ what is happening on GA on a local Intel Mac?

**Reply by joerick** [reply]
On an Intel Mac, you can do `cibuildwheel --platform macos --archs arm64` to try a build locally. As long as you have Command-line Tools For Xcode installed, that will be very similar to the GHA build.

**Reply by lelit** [reply]
Much appreciated, thank you!

**Comment by lelit** [comment]
An update before closing this, for posterity.

First thing I tried has been changing the way the extension gets compiled: instead of building the `libpg_query` static library I explicitly [listed  all its sources](https://github.com/lelit/pglast/blob/issue144/setup.py#L33-L38), and that **worked**, to my surprise, producing a loadable binary also on Silicon macs.

I eventually switched to the new `macos-14` GH Actions runner, re-enabling tests, and as expected that worked too.

Thanks again Joe for your hints!
